packages/yoroi-extension/app/components/wallet/create-wallet/VerifyRecoveryPhraseStep.js

```diff
@@ -0,0 +1,239 @@
+// @flow
+import { useState, useMemo } from 'react';
+import type { Node, ComponentType } from 'react';
+import { defineMessages, injectIntl, FormattedHTMLMessage } from 'react-intl';
+import { observer } from 'mobx-react';
+import type { $npm$ReactIntl$IntlShape } from 'react-intl';
+import { Stack, Box, Typography, Button } from '@mui/material';
+import StepController from './StepController';
+import { CREATE_WALLET_SETPS } from './steps';
+import styles from './VerifyRecoveryPhraseStep.scss';
+import classnames from 'classnames';
+import { ReactComponent as VerifiedIcon } from '../../../assets/images/verify-icon-green.inline.svg';
+import environment from '../../../environment';
+import { makeSortedPhrase } from '../../../utils/recoveryPhrase';
+import globalMessages from '../../../i18n/global-messages';
+
+const messages = defineMessages({
+  description: {
+    id: 'wallet.create.thirdStep.description',
+    defaultMessage:
+      '!!!<strong>Select</strong> each word in <strong>the correct order</strong> to confirm your recovery phrase.',
+  },
+  incorrectOrder: {
+    id: 'wallet.create.thirdStep.incorrectOrder',
+    defaultMessage: '!!!Incorrect order. Try again',
+  },
+  verified: {
+    id: 'walllet.create.thirdStep.verifiedRecoveryPhrase',
+    defaultMessage: '!!!The recovery phrase is verified',
+  },
+});
+
+type Intl = {|
+  intl: $npm$ReactIntl$IntlShape,
+|};
+
+type Props = {|
+  setCurrentStep(stepId: string): void,
+  recoveryPhrase: Array<string> | null,
+|};
+
+function VerifyRecoveryPhraseStep(props: Props & Intl): Node {
+  const { intl, recoveryPhrase, setCurrentStep } = props;
+  if (!recoveryPhrase) throw new Error('Missing recovery phrase, should never happen');
+
+  const [enteredRecoveryPhrase, setRecoveryPhrase] = useState(
+    new Array(recoveryPhrase.length).fill(null)
+  );
+  const [wrongWord, setWrongWord] = useState<string | null>(null);
+
+  function onAddWord(word: string, idx: number): void {
+    if (isWordAdded(word)) return;
+
+    const nextWordIdx = enteredRecoveryPhrase.findIndex(w => w === null);
+    if (nextWordIdx === -1) throw new Error('Entered recovery phrase words list is full');
+
+    const isInCorrectOrder = recoveryPhrase[nextWordIdx] === word;
+    if (!isInCorrectOrder) {
+      return setWrongWord(word);
+    }
+
+    setRecoveryPhrase(prev => {
+      const copy = [...prev];
+      copy[idx] = word;
+      return copy;
+    });
```

Comment:
@AhmedIbrahim336 , in this function you receive parameter `idx`. Later in the code `originalidx` is passed into that parameter. By the logic you are using this parameter as "the index where to add the word", because you are assigning the new recovery phrase as `copy[idx] = word;`. BUT you are using `nextWordIdx` for validation, which is calculated as the first **empty** index of the entered recovery phrase: `enteredRecoveryPhrase.findIndex(w => w === null)`. This creates a weird behaviour case, for example

Imagine you have a recovery phrase `"b a b"`, the `recoveryPhrase` array is then `['b', 'a', 'b']`, in the sorted words the user will see options: 'a', 'b', and 'b', and the user needs to enter the first word of the phrase which is `b`. From the user perspective both words `b` are equally valid, but internally they have different `originalIdx` values: one is `0` and the other is `2`.

In your implementation if the user picks `b` with the `originalIdx = 2` it will pass the validation, because the `nextWordIdx` will be equal to `0` and `recoveryPhrase[0] === 'b'` will pass, but then the `setRecoveryPhrase` will actually assign the selected word to the index `2` instead of the index `0` because the `originalIdx` passed to `onAddWord` will be `2`, so the user tries to enter the first word and will instead get result `[null, null, b]`.

This must be changed. Don't enforce setting the words to the same index as they have been at originally. Logically equal words are interchangeable. The original index must only be used as a unique word identifier, so they are not compared just by text **internally**. I'd recommend to actually abandon storing `originalidx` and instead call it like:

```
internalWordId: `${word}-${originalIdx}`
```

And then only use this field for internal checks that require uniqueness, e.g. 
```
setWrongWord(selectedWord.internalWordId);
```